pkg/scheduler/plugins/elasticquota/core/quota_tree_wrapper.go

```diff
@@ -0,0 +1,501 @@
+}
+
+func NewQuotaNode(quotaName QuotaNameType, sharedRatio, request, min int64, allowLentResource bool) *quotaNode {
+	if sharedRatio == 0 {
+		return nil
+	}
+
+	return &quotaNode{
+		quotaName:         quotaName,
+		request:           request,
+		sharedRatio:       sharedRatio,
+		minQuota:          min,
+		runtimeQuota:      0,
+		allowLentResource: allowLentResource,
+	}
+}
+
+type quotaTree struct {
+	quotaNodes map[QuotaNameType]*quotaNode
+}
+
+func NewQuotaTree() *quotaTree {
+	return &quotaTree{
+		quotaNodes: make(map[QuotaNameType]*quotaNode),
+	}
+}
+
+func (qt *quotaTree) insert(groupName QuotaNameType, sharedRatio, request, min int64, allowLentResource bool) {
+	if _, exist := qt.quotaNodes[groupName]; !exist {
+		qt.quotaNodes[groupName] = NewQuotaNode(groupName, sharedRatio, request, min, allowLentResource)
+	}
+}
+
+func (qt *quotaTree) erase(groupName QuotaNameType) {
+	if _, exist := qt.quotaNodes[groupName]; exist {
+		delete(qt.quotaNodes, groupName)
+	}
+}
+
+func (qt *quotaTree) update(groupName QuotaNameType, sharedRatio, request, min int64, allowLentResource bool) {
+	if nodeValue, exist := qt.quotaNodes[groupName]; exist {
+		if nodeValue.sharedRatio != sharedRatio || nodeValue.request != request || nodeValue.minQuota != min ||
+			nodeValue.allowLentResource != allowLentResource {
+			qt.quotaNodes[groupName] = NewQuotaNode(groupName, sharedRatio, request, min, allowLentResource)
+		}
+	}
+}
+
+func (qt *quotaTree) find(groupName QuotaNameType) (bool, *quotaNode) {
+	if nodeValue, exist := qt.quotaNodes[groupName]; exist {
+		return exist, nodeValue
+	}
+
+	return false, nil
+}
+
+func (qt *quotaTree) adjustQuota(totalResource, totalRequest int64) {
+	if totalRequest < totalResource {
+		qt.adjustQuota4RequestLessEqualResource()
+	}
+	toPartitionResource := totalResource
+	totalSharedRatio := int64(0)
+	needAdjustQuotaNodes := make([]*quotaNode, 0)
+	for _, node := range qt.quotaNodes {
+		runtimeQuotaTmp := node.request
+		if runtimeQuotaTmp > node.minQuota {
+			//if a node's request > min, the node needs adjustQuota
+			runtimeQuotaTmp = node.minQuota
+			needAdjustQuotaNodes = append(needAdjustQuotaNodes, node)
+			totalSharedRatio += node.sharedRatio
+		}
+		if node.allowLentResource {
+			node.runtimeQuota = runtimeQuotaTmp
+		} else {
+			node.runtimeQuota = node.minQuota
+		}
+		toPartitionResource -= node.runtimeQuota
+	}
+	if toPartitionResource > 0 {
+		qt.Iteration4AdjustQuota(toPartitionResource, totalSharedRatio, needAdjustQuotaNodes)
+	}
+}
+
+func (qt *quotaTree) Iteration4AdjustQuota(totalRes, totalSharedRatio int64, nodes []*quotaNode) {
+	needAdjustQuotaNodes := make([]*quotaNode, 0)
+	toPartitionResource, needAdjustTotalSharedRatio := int64(0), int64(0)
+	totalResCopy := totalRes
+	for _, node := range nodes {
+		runtimeQuotaDelta := int64(float64(node.sharedRatio)*float64(totalResCopy)/float64(totalSharedRatio) + 0.5)
+		node.runtimeQuota += runtimeQuotaDelta
+		totalRes -= runtimeQuotaDelta
```

Comment:
totalRes never used in the follow steps, it seems to be totalResCopy?

Reply:
done